### ISSUE TITLE
Node integration

### ISSUE THREAD

**Issue by maierfelix** [gpuweb/cts#117]
Hey,

I'd like to use this for the node bindings - @kainino0x could you take a look into how the cts could be integrated? Need some initial steps to get started

Thanks

**Comment by kainino0x** [comment]
Hey Felix, unfortunately this can't be a priority for me so it'll be at least 2 weeks before I can consider it. I think it will be pretty quick for you to do it - and easier, since you already have builds.

All you need to do is get the getGPU() function to return an object that implements requestAdapter:
https://github.com/gpuweb/cts/blob/master/src/framework/gpu/implementation.ts

I used to have support in there for my proof-of-concept node bindings (that had like 5 entry points implemented). You can look at how it was done then as an example:
https://github.com/gpuweb/cts/blob/26b96dac48eeaee39ccb62c905b34f59cfd74129/src/framework/gpu/implementation.ts

**Comment by maierfelix** [comment]
Thanks for the useful information, I will try

**Comment by dcerisano** [comment]
Hey guys, I am working on integrating Kai's `CTS` and Felix's `node-webgpu`.
I see that the official webgpu spec is imported into `CTS` from the `@webgpu `package.

Seems that `@webgpu` is currently designed for chromium and so does not provide for the stand-alone (addon) requirement for a window, since chromium provides one internally.

Felix extended the addon API by providing a WebGPUWindow.
He also changed the GPU.requestAdapter API by passing a GPUWindow there instead of GPURequestAdpaterOptions as specified by @webgpu.

So I think we need to add GPUWindow to the @webgpu spec and also add WebGPUWindow as in GPURequestAdapterOptions. That would be one way of satisfying the requirement for both chromium and stand-alone modes in the official @webgpu API.

Basically just need a way to specify the window context if none exists. Again in chromium of course one always exists.

Could also just overload the requestAdapter function to do the window creation internally, I suppose.

My test branch where I am working on this is [here](https://github.com/dcerisano/cts/tree/test-branch), a clone of Kai's experimental tree he gave above.

**Comment by dcerisano** [comment]
Hey guys, I am working on integrating Kai's CTS and Felix's node-webgpu while Felix is completing his super awesome node-webgpu RT demos.

I see that the official webgpu spec is imported into CTS from the @webgpu package. Seems that @webgpu is currently designed mostly for browser support and so does not provide for the stand-alone (eg. addon) requirement to set a window, since browsers provide one internally.

Felix extended the addon API by providing a WebGPUWindow. He also changed the addon GPU.requestAdapter API by passing a WebGPUWindow there instead of GPURequestAdpaterOptionsas specified by @webgpu.

So I think we could add GPUWindow to the @webgpu spec and also add GPUWindow in GPURequestAdapterOptions so we can keep the same signature. That would be one way of satisfying the requirement for both browser and stand-alone modes in the official @webgpu API.

I have seen on other APIs the ability to set a Graphics Context (GC), which is what a GPUWindow is. Maybe call it GPUContext instead of GPUWindow? It would be kind of neat if it could just be a canvas or other lower level GC, allowing for WebGPU powered desktops, overlays, screensavers, widgets, wallpapers, etc.

My main focus is GPU compute, which of course does not require a GC at all. So I guess maybe @webgpu API could allow for no GC (headless) mode too?

My test branch where I am working on this is here, a clone of Kai's experimental tree he gave above. Please note I am currently focused on the linux platform and will try to keep any PRs I submit cross-platform. The changes currently in there are certainly not as 'typescripty' as they should be.

Pardon the delay getting going with this, Typescript was a surprise to me, and slowly beginning to appreciate its value, especially in a complex API like @webgpu. It already caught this issue, and kind of led to a solution.

**Comment by kainino0x** [comment]
What package are you talking about by `@webgpu`? Do you mean `@webgpu/types`? The WebGPU spec is not imported into the CTS - there is no JS code in it.

I don't see how spec changes could make the same code work in both. You'll still need to do, essentially, `const gpu = (typeof navigator !== 'undefined') ? navigator.gpu : require('webgpu')`.

Re: `GPUWindow`. I don't understand what a GPUWindow is. Is it an OS window-system window? We do not want a window or canvas to create a WebGPU device. WebGPU devices should always be created headless - so they can be used with zero, one, or more canvas/window outputs, via `GPUSwapChain`. I think this should be possible in native for the modern APIs (though OpenGL needs a window to be initialized...).

The native [`webgpu.h`](https://github.com/webgpu-native/webgpu-headers/blob/master/webgpu.h) (used for wgpu-native, Dawn, and Emscripten) defines the concept of a "surface", which can be used to create a swap chain. `GPUSurface` does not exist in the JS API, but it is closely related to `GPUCanvasContext`. webgpu-node SHOULD use the GPUSurface API to hook windows into Dawn, however, we have not finished implementing it yet.

For creating windows, webgpu-node should provide a special extra entry point (since you cannot create `<canvas>` elements), e.g. `gpu.createWindow`, which creates an OS window and returns an object that implements the `GPUCanvasContext` interface (`configureSwapChain` and `getSwapChainPreferredFormat`).

---

P.S. Please be careful not to `@` the github user "`@webgpu`" - that's a project unrelated to W3C WebGPU.

**Comment by dcerisano** [comment]
@webgpu/types is the Typescript definition file (idl) where the webgpu API is defined.

I am seeking to integrate CTS with NAPI, not a browser.

There still needs to be a way to specify a graphics context for non browser applications.

I will look at implementing GPUSurface, thanks - I think that is what I was looking for. 

Node-webgpu already provides gpu.WebGPUWindow but that is not part of the spec and there is not anyway to pass it to the API without changing the spec, which of course @webgpu/types will not permit.

**Comment by kainino0x** [comment]
> @webgpu/types contains the Typescript definition file (index.d.ts) where the official (@) webgpu API is defined. Seems to be the whole point of using Typescript in CTS.

That's largely the point, but also for internal type checking.

> I am seeking to integrate CTS with NAPI, not a browser. CTS seems to have been created for browser validation only, and may require changes to validate non browser applications.

It was created with Node in mind, but it may have broken very slightly since then. We do compile and run the unit tests in Node on CI.

> I will look at implementing GPUSurface, thanks - I think that is what I was looking for. Again pardon my unfamiliarity with the webgpu API which I appreciate is still a WIP. I will raise the issue in group.

I don't think there is any issue that needs to be raised in the group. The gpuweb group is not adding a GPUSurface concept. If you have any webgpu.h related issues they should go here: https://github.com/webgpu-native/webgpu-headers

> Node-webgpu already provides gpu.WebGPUWindow but that is not part of the spec and there is not any way to pass it to the API. Typescript will not permit this without changing @webgpu/types.

You should provide a custom type for `gpu` and `gpu.WebGPUWindow` but use the `@webgpu/types` type definition for `GPUAdapter` for the return type of `gpu.requestAdapter` even if your `requestAdapter` has a different signature. That way you can customize just the part you need and use the upstream types for everything else.

---

Sorry, missed a response before:

> My test branch where I am working on this is [here](https://github.com/dcerisano/cts/tree/test-branch), a clone of Kai's experimental tree he gave above. Please note I am currently focused on the linux platform and will try to keep any PRs I submit cross-platform. The changes currently in there are certainly not as 'typescripty' as they should be.

That's not an experimental branch, it's a very old revision. Please update to the latest version of the CTS but use that old code as inspiration.

**Comment by dcerisano** [comment]
> That's largely the point, but also for internal type checking.
Type checking is the whole point of Typescript. There is no other point to it. In this application, compliance with the official @webgpu API spec, which is is only defined in Typescript, AFAIK.

> It was created with Node in mind, but it may have broken very slightly since then. We do compile and run the unit tests in Node on CI.

I am integrating against NAPI, not necessarily Node. My ultimate goal is to integrate Felix's NAPI webgpu library directly with V8, not Node ("because security reasons").

> You should provide a custom type for gpu and gpu.WebGPUWindow but use the @webgpu/types type definition for GPUAdapter for the return type of gpu.requestAdapter even if your requestAdapter has a different signature. That way you can customize just the part you need and use the upstream types for everything else.

It seems (again pardon my relative newness) that webgpu API can be used without upstream modification in-browser only. There does seem to be a requirement to include any web-enabled application (eg. Node, direct V8, etc). I will raise it in group, but I appreciate most (all?) of the stakeholders are browsers and this is not a priority.

> That's not an experimental branch, it's a very old revision. Please update to the latest version of the CTS but use that old code as inspiration.

The source tree you provided above (which my test-branch is cloned from) is marked as experimental in it's README. At this point, I am just trying to get a single CTS unit test to run with NAPI before even thinking about any PRs.

**Comment by kainino0x** [comment]
> Type checking is the whole point of Typescript. There is no other point to it. In this application, compliance with the official @webgpu API spec, which is is only defined in Typescript, AFAIK.

I'm saying that it's to check two sets of types: the `@webgpu/types` types for the API, and *all of the types inside the CTS itself*. The CTS has a considerable amount of framework code and the types massively help write correct framework code and test instantiations.



> I am integrating against NAPI, not necessarily Node. My ultimate goal is to integrate Felix's NAPI webgpu library directly with V8, not Node ("because security reasons").

Good point, thanks for clarifying. That said, I don't think it will be any more difficult with V8 than with Node.

> It seems (again pardon my relative newness) that webgpu API can be used without upstream modification in-browser only. There does seem to be a requirement to include any web-enabled application (eg. Node, direct V8, etc). I will raise it in group, but I appreciate most (all?) of the stakeholders are browsers and this is not a priority.

Apps that use webgpu-napi can use a .d.ts file that's packaged along with webgpu-napi, and that .d.ts can have a dependency on the upstream `@webgpu/types` .d.ts file.

Also note that `@webgpu/types` is not really maintained by the WebGPU group, it's maintained by me.

> The source tree you provided above (which my test-branch is cloned from) is headlined as "# cts-experiment" in it's README.

That's because, at that time, we had not yet made the CTS "official". It was my "experiment" in creating the very first CTS for WebGPU.

> Also, it seems to have been updated only a few months ago, which is not "very old" or even "old" by any stretch. "Stale" might be more appropriate, but it was the only reference you gave in this thread.

It is 7.5 months old, and a lot has changed since then. There were a miniscule number of tests at that time - almost all of the tests were added since then. The API surface of WebGPU itself has changed considerably - if webgpu-napi matches the latest webgpu.dev, it's not going to be compatible with that version of the CTS.

I didn't reference "master of the repository this issue is on" because it didn't seem necessary.

**Comment by dcerisano** [comment]
> I didn't reference "master of the repository this issue is on" because it didn't seem necessary.

It was necessary for you to reference the experimental source tree because the master branch does not provide the framework in implementation.ts for testing against an out-of-browser application. Only the experimental branch attempts this.

Curious, do you have a reference for the index.node addon you reference in the experimental source tree? I am using Felix's NAPI addon in it's place and am wondering how you got CTS to work with an addon.

**Comment by kainino0x** [comment]
Yes, because I deleted that support. However it should be nearly trivial to reintroduce that support in implementation.ts, and this is the necessary path forward.

> Curious, do you have a reference for the index.node addon you reference in the experimental source tree? I am using Felix's NAPI addon in it's place and am wondering how you got CTS to work with your addon.

Excellent question. Here's the abandoned patch where the code lives that created that addon.
https://dawn-review.googlesource.com/c/dawn/+/3640
It only had a few entry points.
I don't remember how this stuff works. I hope this is a helpful pointer.

**Comment by dcerisano** [comment]
Hey @maierfelix ! Kai said "trivial"! LOL. Felix nearly bit my head off when I said that. 

Certainly it will not be much code in the end (because I am determined to use Felix's NAPI addon - the thing that is actually being tested), but the complexity of the WIPs involved is staggering (even @kainino0x does not remember "how this stuff works" and he wrote it) on top of that the thing has to be written in Typescript.

Thanks very much for this discussion. I now understand why I feel like a Marine who parachuted into a jungle alone on this. Going to start hacking my way out, hopefully my buddies will come choppering in at some point.

![image](https://user-images.githubusercontent.com/1641613/74383267-e2a6e400-4dbc-11ea-8193-0193e543278b.png)

**Comment by kainino0x** [comment]
I'm sorry to have implied the work is trivial. Obviously it is not. I am saying only that the amount of code that was removed from the CTS is nearly trivial and can be easily re-added by "copying" from the old code to the new code. Actually making the integration work (getting the functions to match, the API to match, etc.) is not easy.

**Comment by dcerisano** [comment]
I have done a preliminary [integration](https://github.com/dcerisano/cts) of Felix's webgpu addon and the CTS head.

I am getting these [errors](https://gist.github.com/dcerisano/58b733478c3d1d6e0edd46eecff1f4dd).

The problem seems to be that CTS has overridden nodejs filename resolution for internal require() statements. This generates a special listing to work around browser file system limitations.

The override will not handle addons (index.node) files, which are loaded via the require() statement. I am not running within a browser, so how do I turn off the listing generator and/or nodejs override?

**Comment by kainino0x** [comment]
Sorry I haven't had a chance to look into this today. Sorry about the weird problem with overriding require. It is not for the browser, it is actually node specific. It allows us to write the command line tools in TypeScript and share code with the rest of the codebase.

Is the issue that it wants all files to end in .js? Maybe you can just also allow .node here: https://github.com/gpuweb/cts/blob/master/src/tools/setup-ts-in-node.js#L16
Otherwise let me know and I will take a closer look.

On including between the cts and the implementation, it would make most sense to include the cts from the implementation. But I will have to think about how to do. I don't think we'll publish the cts as a npm package, but maybe it would work to call into it as a submodule.

**Comment by dcerisano** [comment]
Completed the draft integration of CTS and NAPI WebGPU addon [here](https://github.com/dcerisano/cts).

Hitting a new (and better) issue now.
When I `tools/run cts` my desktop freezes for a minute while 790 windows are created for each CTS test and then a core dump:

```
Warning: terminator_CreateDevice: Failed in ICD libGLX_nvidia.so.0 vkCreateDevicecall
Warning: vkCreateDevice:  Failed to create device chain.
Info: vkCreateDevice failed with VK_ERROR_INITIALIZATION_FAILED
Segmentation fault (core dumped)
```

I am assuming this is because I am simply running out of memory. Again pardon my newness to CTS, but is there a way to easily run the tests sequentially rather than in parallel?

**Comment by kainino0x** [comment]
Hm, sorry about this. I think I still have a TODO to not run all the tests at once. If that's still there I can fix it soon.
That said, we also definitely need to get those windows to close (or at least create fewer of them). We need to make sure there's a way to close the windows reliably without relying on the garbage collector (so the napi plugin will need to expose a method for that).

Does it currently create one window for every device? I currently create one device for every test case and I have another TODO to not do that.

No worries about any breaking changes for the Web version. I'm happy to help fix anything up as needed when we're merging any code changes you have.

**Comment by dcerisano** [comment]
I can get around it for now using your test filters eg:
`tools/run cts:buffers`

This creates a manageable number of windows, but now getting errors from NAPI concerning error handling:
```
FATAL ERROR: Error::New napi_get_last_error_info
 1: 0x9dab70 node::Abort() [node]
 2: 0x9dbd26 node::OnFatalError(char const*, char const*) [node]
 3: 0x9dbe19  [node]
 4: 0x9aed9b napi_fatal_error [node]
 5: 0x7fe824c897ed  [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
 6: 0x7fe824c89734 Napi::Error::New(napi_env__*) [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
 7: 0x7fe824d08d1c  [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
 8: 0x7fe82c09b0fa  [/home/dcerisano/git/webgpu/generated/0.0.1/linux/build/Release/libdawn_native.so]
 9: 0x7fe824cf5b90 GPUBuffer::mapReadAsync(Napi::CallbackInfo const&) [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
10: 0x7fe824d06a42 GPUDevice::createBufferMapped(Napi::CallbackInfo const&) [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
11: 0x7fe824d0992b Napi::ObjectWrap<GPUDevice>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*)::{lambda()#1}::operator()() const [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
12: 0x7fe824d0985a Napi::ObjectWrap<GPUDevice>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*) [/home/dcerisano/git/cts/third_party/dawn/linux/index.node]
13: 0x98f765  [node]
14: 0xba6fc9  [node]
15: 0xba8db7 v8::internal::Builtin_HandleApiCall(int, unsigned long*, v8::internal::Isolate*) [node]
16: 0x1376359  [node]
Aborted (core dumped)
```

So next request is for both @kainino0x  and @maierfelix. Could you please review my [implementation](https://github.com/dcerisano/cts/blob/master/src/framework/gpu/implementation.ts) ? The issue is typescript and/or the way I am handling the integration of NAPI.

**Comment by dcerisano** [comment]
I am using the test suite filtering provided by CTS to avoid running all 790 test at once.
eg. `tools/run cts:buffers` produces a managable number tests.

Still having issues, looking into making sure each test exits gracefully (ie. closes it's GPU resources) before running the next. Again concerned this will conflict with existing in-browser testing.

**Comment by dcerisano** [comment]
Still poking away at this. I got one test suite to pass:
```
dcerisano@kronos:~/git/cts$ tools/run cts:canvas

** Summary **
Passed  w/o warnings = 1 / 1 = 100.00%
Passed with warnings = 0 / 1 =   0.00%
Failed               = 0 / 1 =   0.00%
```
The other test suites involve more than one test, and core dump (see above).

@kainino0x indicated he has some TODOs that should allow running these suites. 

I would like to create a PR,  and then merge back the TODOs once @kainino0x has them. I think we can get some more suites to pass this way.

**Comment by dcerisano** [comment]
Still poking away at this. I got one test to pass.

```
dcerisano@kronos:~/git/cts$ tools/run cts:canvas

** Summary **
Passed  w/o warnings = 1 / 1 = 100.00%
Passed with warnings = 0 / 1 =   0.00%
Failed               = 0 / 1 =   0.00%

```

It should not have passed, since it is browser based, and I am testing a NAPI addon, not in browser.

@kainino0x could you create a new branch `cts-napi` with those TODOs you mentioned? I still can't run any suite that has more than one test.  I would like to submit PRs against that branch for now, not the master.

**Comment by kainino0x** [comment]
Sorry for the delay. I'm looking at putting some fixes in now.

**Comment by kainino0x** [comment]
> It should not have passed, since it is browser based, and I am testing a NAPI addon, not in browser.

It is actually supposed to pass. It checks for the existence of `document`, in order to figure out whether it's running in a browser main thread. If `document` doesn't exist, then it skips itself (which is a "pass"). This is so that it doesn't meaninglessly fail on Web Workers and in Node.
https://github.com/gpuweb/cts/blob/master/src/suites/cts/canvas/context_creation.spec.ts#L8

**Comment by dcerisano** [comment]
Is there a way to turn on messages like:
`t.skip('DOM is not available to create canvas element');`

Really I would like to turn on all the messaging there is - this is proving very challenging to debug.

**Comment by kainino0x** [comment]
I don't think there is a way to do that right now in the cmdline runner, but there is a "debug" mode that is wired up elsewhere. I will wire it up into the cmdline runner.

**Comment by kainino0x** [comment]
Actually, the problem was I didn't make the cmdline runner understand the "skip" status (oops). Fixing that and wiring up the debug mode.

**Comment by kainino0x** [comment]
Ok, done and merged. Once you update, try running tools/run with `--verbose --debug`. (Though `--debug` will probably not be that helpful most of the time.) This should be much better for debugging.

I created a new cts-napi branch which currently is the same as master. I'm happy to take PRs in there anytime and I'll take care of any issues (e.g. in case there are any issues with running in browser) before merging that branch to master.

**Comment by kainino0x** [comment]
And, of course, let me know if you run into any more issues (or if any of the stuff I just did isn't working - I didn't test it exhaustively).

**Comment by dcerisano** [comment]
Thanks @kainino0x, no problem.
I have a feeling I will be testing it exhaustively :)

**Comment by dcerisano** [comment]
Just got back to this. Thanks for the last push.

I am getting some of the CTS tests to pass, but not happy with the current approach.
The issue right now is that the WebGPU addon is not fully compliant with the @webgpu spec.

Rather than bending @webgpu to fit the N-API (current approach), I am thinking the making the N-API fit @webgpu. It is going to have to be done anyway as the official spec evolves.

This will mean PRs to @maierfelix's WebGPU addon in addition to PR's here. I will ask @maierfelix  to create a cts branch in his repo.

**Comment by kainino0x** [comment]
Sounds like a good plan! Thanks for the update!

**Comment by Kangz** [comment]
Closing this. The CTS runs well inside [dawn_node](https://dawn.googlesource.com/dawn/+/refs/heads/main/src/dawn_node/).
